File: src/srf_punctuation/inference.py
import json
from pathlib import Path
from typing import Dict, List

import torch

from .config import Config
from .models import PunctuationPredictor


class PunctuationInference:
    def __init__(
        self,
        model_path: str,
        vocab_path: str,
        config: Config = None,
    ):
        self.config = config or Config()
        self.vocab: Dict[str, int] = {}
        self.id_to_char: Dict[int, str] = {}
        self._load_vocab(vocab_path)
        self._load_model(model_path)

    def _load_vocab(self, vocab_path: str) -> None:
        with open(vocab_path, "r", encoding="utf-8") as f:
            self.vocab = json.load(f)
        self.id_to_char = {v: k for k, v in self.vocab.items()}

    def _load_model(self, model_path: str) -> None:
        checkpoint = torch.load(model_path, map_location="cpu", weights_only=False)
        self.model = PunctuationPredictor(
            vocab_size=self.config.model.vocab_size,
            embed_dim=self.config.model.embed_dim,
            hidden_dim=self.config.model.hidden_dim,
            num_layers=self.config.model.num_layers,
            num_heads=self.config.model.num_heads,
            num_labels=self.config.model.num_labels,
            dropout=0.0,
            max_seq_len=self.config.model.max_seq_len,
        )
        if "state_dict" in checkpoint:
            state_dict = checkpoint["state_dict"]
            model_state = {}
            for k, v in state_dict.items():
                if k.startswith("model."):
                    model_state[k[6:]] = v
            self.model.load_state_dict(model_state)
        else:
            self.model.load_state_dict(checkpoint)
        self.model.eval()

    def text_to_ids(self, text: str) -> List[int]:
        return [self.vocab.get(char, self.vocab["<UNK>"]) for char in text]

    def predict(self, text: str) -> str:
        char_ids = self.text_to_ids(text)
        if len(char_ids) > self.config.model.max_seq_len:
            char_ids = char_ids[: self.config.model.max_seq_len]

        input_ids = torch.tensor([char_ids], dtype=torch.long)
        attention_mask = torch.ones(1, len(char_ids), dtype=torch.long)

        with torch.no_grad():
            outputs = self.model(input_ids, attention_mask)
            logits = outputs["logits"]
            preds = torch.argmax(logits, dim=-1)[0].tolist()

        punctuation_tokens = self.config.punctuation_tokens
        result = []
        for i, char in enumerate(text[: len(preds)]):
            result.append(char)
            label = preds[i]
            for name, idx in self.config.punctuation_map.items():
                if idx == label and name != "O":
                    result.append(punctuation_tokens[name])
                    break

        return "".join(result)

    def predict_batch(self, texts: List[str]) -> List[str]:
        return [self.predict(text) for text in texts]


def load_inference(
    checkpoint_path: str = "checkpoints/punctuation-epoch=01-val_loss=0.6772.ckpt",
    vocab_path: str = "data/vocab.json",
) -> PunctuationInference:
    return PunctuationInference(checkpoint_path, vocab_path)

File: web/src/inference.ts
import { Config, Vocab, PunctuationMap } from './types'

const defaultConfig: Config = {
  model: {
    vocabSize: 8000,
    embedDim: 128,
    hiddenDim: 256,
    numLayers: 4,
    numHeads: 8,
    numLabels: 5,
    maxSeqLen: 256
  },
  punctuationMap: {
    O: 0,
    COMMA: 1,
    PERIOD: 2,
    QUESTION: 3,
    EXCLAMATION: 4
  },
  punctuationTokens: {
    O: '',
    COMMA: '，',
    PERIOD: '。',
    QUESTION: '？',
    EXCLAMATION: '！'
  },
  chinesePunctuation: ['，', '。', '？', '！', '、', '；', '：', '\u201c', '\u201d', '\u2018', '\u2019', '（', '）', '【', '】', '《', '》', ',', '!', '?', '.', ';', ':', '(', ')', '[', ']']
}

export class PunctuationInference {
  private session: Awaited<ReturnType<typeof import('onnxruntime-web').InferenceSession.create>> | null = null
  private vocab: Vocab = {}
  private config: Config = defaultConfig
  private loaded = false

  async loadModel(modelPath: string, vocabPath: string): Promise<void> {
    const ort = await import('onnxruntime-web')

    const [session, vocabResp] = await Promise.all([
      ort.InferenceSession.create(modelPath, {
        executionProviders: ['wasm']
      }),
      fetch(vocabPath).then(r => r.json()) as Promise<Vocab>
    ])
    
    this.session = session
    this.vocab = vocabResp
    this.loaded = true
  }

  private cleanText(text: string): string {
    return text.split('').filter(c => !this.config.chinesePunctuation.includes(c)).join('')
  }

  private textToIds(text: string): number[] {
    const cleanText = this.cleanText(text)
    const unkId = this.vocab['<UNK>'] ?? 1
    return cleanText.split('').map(char => this.vocab[char] ?? unkId)
  }

  isLoaded(): boolean {
    return this.loaded
  }

  async predict(text: string): Promise<string> {
    if (!this.session) {
      throw new Error('Model not loaded')
    }

    const ort = await import('onnxruntime-web')
    const cleanText = this.cleanText(text)
    const charIds = this.textToIds(text)
    
    if (charIds.length > this.config.model.maxSeqLen) {
      charIds.splice(this.config.model.maxSeqLen)
    }

    const inputIds = new ort.Tensor('int64', BigInt64Array.from(charIds.map(BigInt)), [1, charIds.length])
    const attentionMask = new ort.Tensor('int64', BigInt64Array.from(new Array(charIds.length).fill(BigInt(1))), [1, charIds.length])

    const results = await this.session.run({
      input_ids: inputIds,
      attention_mask: attentionMask
    })

    const logits = results.logits.data as Float32Array
    const preds: number[] = []
    
    for (let i = 0; i < charIds.length; i++) {
      let maxIdx = 0
      let maxVal = -Infinity
      for (let j = 0; j < this.config.model.numLabels; j++) {
        const val = logits[i * this.config.model.numLabels + j]
        if (val > maxVal) {
          maxVal = val
          maxIdx = j
        }
      }
      preds.push(maxIdx)
    }

    return this.decode(cleanText.slice(0, preds.length), preds)
  }

  private decode(text: string, preds: number[]): string {
    const punctuationTokens = this.config.punctuationTokens
    const punctuationMap = this.config.punctuationMap
    
    const reverseMap: Record<number, keyof PunctuationMap> = {}
    for (const [key, value] of Object.entries(punctuationMap) as [keyof PunctuationMap, number][]) {
      reverseMap[value] = key
    }

    const result: string[] = []
    for (let i = 0; i < text.length; i++) {
      result.push(text[i])
      const label = preds[i]
      const punctName = reverseMap[label]
      if (punctName && punctName !== 'O') {
        result.push(punctuationTokens[punctName])
      }
    }

    return result.join('')
  }
}

export const inference = new PunctuationInference()